napari/layers/image/image_interface.py

```diff
@@ -0,0 +1,36 @@
+from abc import abstractmethod
+from ..base._base_layer_interface import BaseLayerInterface
+
+
+class ImageLayerInterface(BaseLayerInterface):
+    """
+    Defines the getters/setters editable across components for an ImageLayer
+    """
+
+    @abstractmethod
+    def _on_interpolation_change(self, value):
```

Comment:
still wondering about the need for this interface... do these serve any other purpose than to make sure that the `getattr` call in `EventHander.on_change` doesn't throw an error?  And if so, can't we just use a `try/except` or a default to `getattr`?
```python
method = getattr(component, update_method_name, None)
if method:
    ...
```

Reply:
This can be checked 'at compile time' which with python means that for example pycharm can tell you right away that you have non-compliant implementation even before you run the code. Also, interfaces are implicitly  documentation about which methods need to be implemented. it might be difficult to execute all code paths to figure out that 'all is there'. of course, tests can do a lot to help in this respect, but it is good to have this explicit declaration that can be checked 'offline'.

Reply:
I do think simply having things explicitly documented and discoverable by IDEs is sufficient... and we discussed that too in the "atlantic" meeting yesterday.  But this is not exactly a traditional python interface (it doesn't impose anything on its subclasses because it doesn't inherit from ABC, and it is not a `typing.Protocol`), and while I too lean towards using it, I think it's important that we understand exactly _why_ it's there.
Just to play devils advocate for a moment, all of these `_on_*_change` actually _are_ defined elsewhere, so IDEs will be able to find them... but also, internally, in this PR, these methods are mostly being called in `LayerEventHandler.on_change`, where a `getattr(str)`  is already being used to look them up (which pycharm can't trace anyway)... so while what you're saying here is _usually_ the case for an interface like this... my point was that in this case, it's _kinda_ not.  (but again, simply having one place to document the "intentions" for each event is perhaps justification enough)

Reply:
Ok so I'm a little in the middle here. I like the self-documenting aspect of having the interface, but also it's ugly. The way I would *want* the interface documented is something like:

```python
@dataclass
class ImageLayerInterface:
    interpolation : str = 'nearest'
    contrast_limits : Tuple[int] = (0, 1)
    rendering : str = 'mip'
    iso_threshold : float = 0.5
    attenuation : float = 1.0
    gamma : float = 1.0
    colormap : str = 'gray'
```

That's of course incorrect, but perhaps there's a class decorator we could come up with that would generate the abstract interface for us from a code snippet like the above?

Reply:
And btw, at a minimum, my comment above in the BaseLayerInterface applies: the methods should at least have docstrings describing what they're about.

Reply:
while this looks a lot cleaner, I fear that it does not make it obvious that interfaces need to implement a `_on_[name]_change` method. also, due to the autogenerated nature of these classes, they won't work with PyCharm if we're checking using abstract methods, but it may work if it creates a `typing.Protocol` since I believe that PyCharm uses mypy behind the scenes, but I'd need to check it to find out

Reply:
Having now tried to get tests working with https://github.com/shanaxel42/napari/pull/2 I'm leaning away from this as we ended up needing to register components that don't implement the full interface (see QtLayerWidget) and I think there's nothing wrong with that. Given that this is all internally facing code as well (i.e. no plugin developer would ever be adding an interface here) I really think we can do without them.

Reply:
I went with the method proposed here https://github.com/napari/napari/pull/1040/files#r393260527